docs(oss): overhaul deepagents skills documentation (#4271)
langchain-ai/docs: src/oss/deepagents/skills.mdx

import BackendReadonlySkillsPy from '/snippets/code-samples/backend-readonly-skills-py.mdx';
import BackendReadonlySkillsJs from '/snippets/code-samples/backend-readonly-skills-js.mdx';

As agents take on more complex tasks, the context they need grows with them. Loading all instructions into the system prompt wastes tokens on information irrelevant to the current task, and providing the same guidance manually across sessions does not scale.

Skills solve this by packaging domain expertise, such as workflows, best practices, scripts, reference docs, and templates, into reusable directories. The agent gets a summary of the contents on startup and discovers and reads the contained files only when relevant.

:::js
<Note>
Skills require `deepagents>=1.7.0`.
</Note>
:::

> Deep agent skills follow the [Agent Skills specification](https://agentskills.io/specification). For ready-to-use skills that improve your agent's performance on LangChain ecosystem tasks, see the [LangChain Skills](https://github.com/langchain-ai/langchain-skills) repository.

Reviewer: Maybe let's call these out a bit more
```suggestion
Deep agent skills follow the [Agent Skills specification](https://agentskills.io/specification).

<Tip>
For ready-to-use skills that improve your agent's performance on LangChain ecosystem tasks, see the [LangChain Skills](https://github.com/langchain-ai/langchain-skills) repository.
</Tip>
```

import BackendReadonlySkillsPy from '/snippets/code-samples/backend-readonly-skills-py.mdx';
import BackendReadonlySkillsJs from '/snippets/code-samples/backend-readonly-skills-js.mdx';

As agents take on more complex tasks, the context they need grows with them. Loading all instructions into the system prompt wastes tokens on information irrelevant to the current task, and providing the same guidance manually across sessions does not scale.

> Skills solve this by packaging domain expertise—workflows, best practices, scripts, reference docs, and templates—into reusable directories that the agent discovers and reads only when relevant.

Reviewer: well the agent knows about the directories, it's the extra context agents discover and read when needed. Maybe:
```suggestion
Skills solve this by packaging domain expertise, such as workflows, best practices, scripts, reference docs, and templates, into reusable directories. The agent gets a summary of the contents on startup and discovers and reads the contained files only when relevant.
```

Author: Updated.

:::js
<Note>
Skills require `deepagents>=1.7.0`.
</Note>
:::

Deep agent skills follow the [Agent Skills specification](https://agentskills.io/specification). For ready-to-use skills that improve your agent's performance on LangChain ecosystem tasks, see the [LangChain Skills](https://github.com/langchain-ai/langchain-skills) repository.

## What are skills

Each skill is a directory containing a `SKILL.md` file: a markdown file with YAML frontmatter (`name` and `description`) followed by instructions the agent follows when the skill is activated. A skill directory can also include supporting files:

> - **`SKILL.md`** (required): Metadata and instructions for the agent

Reviewer: well this now shouldn't be in the list, because it's not one of the supporting files

Author: Updated.

1. **Discovery**: At agent start, the middleware scans the configured skill paths, parses each `SKILL.md` frontmatter, and injects skill names and descriptions into the system prompt.
2. **Read**: When the agent determines a skill matches the current task, it reads the full `SKILL.md` content via `read_file`. There is no dedicated activation mechanism.
3. **Execute**: The agent follows the skill's instructions and reads any supporting files (scripts, references, assets) as needed.

The [Agent Skills specification](https://agentskills.io/specification) recommends ~100 tokens for frontmatter and under 5,000 tokens for the `SKILL.md` body. Following these guidelines matters because every skill's frontmatter is added to the system prompt at discovery, while the full body is only read when activated. Keeping both layers small means you can load many skills without crowding the context window.

<Tip>
    Write clear, specific descriptions in your `SKILL.md` frontmatter. The agent decides whether to activate a skill based on the description alone. Include specific keywords and use cases so the agent can match accurately.
</Tip>

> ## Examples

Reviewer: It might make sense to put examples last on this page and link to them from the usage/getting started section

Author: Updated.

1. **Discovery**: At agent start, the middleware scans the configured skill paths, parses each `SKILL.md` frontmatter, and injects skill names and descriptions into the system prompt.
2. **Read**: When the agent determines a skill matches the current task, it reads the full `SKILL.md` content via `read_file`. There is no dedicated activation mechanism.
3. **Execute**: The agent follows the skill's instructions and reads any supporting files (scripts, references, assets) as needed.

The [Agent Skills specification](https://agentskills.io/specification) recommends keeping frontmatter concise and the `SKILL.md` body under 5,000 tokens. Following these guidelines matters because every skill's frontmatter is added to the system prompt at discovery, while the full body is only read when activated. Keeping both layers small means you can load many skills without crowding the context window.

<Tip>
    Write clear, specific descriptions in your `SKILL.md` frontmatter. The agent decides whether to activate a skill based on the description alone. Include specific keywords and use cases so the agent can match accurately.
</Tip>

## Get started

> Create a skill directory with a `SKILL.md` file, then pass the path when creating your agent:

Reviewer: [clarity] want to ensure people know it's a folder with a subfolder
```suggestion
Create a top level skills directory. Then create a directory inside that with a `SKILL.md` file for your first skill. Finally, pass the path to the top level skills directory when creating your agent:
```

:::python
```python
from deepagents import create_deep_agent
from deepagents.backends.filesystem import FilesystemBackend

backend = FilesystemBackend(root_dir="./my-project")

agent = create_deep_agent(
    model="anthropic:claude-sonnet-4-6",
    backend=backend,
    skills=["./my-project/skills/"],
)

result = agent.invoke(
    {"messages": [{"role": "user", "content": "What is LangGraph?"}]},
    config={"configurable": {"thread_id": "1"}},
)
```
:::

:::js
```typescript
import { createDeepAgent, FilesystemBackend } from "deepagents";

const backend = new FilesystemBackend({ rootDir: process.cwd() });

const agent = await createDeepAgent({
  model: "anthropic:claude-sonnet-4-6",
  backend,
  skills: ["/skills/"],
});

const result = await agent.invoke(
  { messages: [{ role: "user", content: "What is LangGraph?" }] },
  { configurable: { thread_id: "1" } },
);
```
:::

This example uses `FilesystemBackend` to load skills from disk. For other storage options, including loading skills from remote sources, see [Backends and remote skill loading](#backends-and-remote-skill-loading).

## Write effective skills

> The [Agent Skills specification](https://agentskills.io/specification) includes guidance on structuring skills for reliable discovery and activation. The recommendations below build on that foundation with practical patterns for Deep Agents.

Reviewer: [style]
```suggestion
The [Agent Skills specification](https://agentskills.io/specification) includes guidance on structuring skills for reliable discovery and activation. The following recommendations build on that foundation with practical patterns for Deep Agents.
```

        ├── schema-reference.md
        └── error-codes.md
```

The agent loads reference files only when the instructions call for them, keeping each layer of progressive disclosure appropriately sized. Keep file references one level deep from `SKILL.md` and avoid deeply nested reference chains, which force the agent through multiple reads to reach the information it needs.

**Structure instructions for the agent.** Write your `SKILL.md` body as clear instructions the agent can follow:

- **Step-by-step procedures** for multi-step workflows
- **Decision criteria** for choosing between approaches
- **Examples of expected inputs and outputs** so the agent knows what success looks like
- **Edge cases** the agent should handle or flag to the user

**Manage skill count.** Fewer well-scoped skills outperform many overlapping ones. As the number of skills with similar descriptions grows, the agent's ability to select the right one degrades. If you find yourself with many related skills, consider:

- Consolidating related capabilities into a single skill with sections for each sub-task
- Using reference files to keep the main `SKILL.md` concise while covering multiple sub-tasks

> <Tip>

Reviewer: that's so cool!

    Use the [`skills-ref` validation tool](https://github.com/agentskills/agentskills/tree/main/skills-ref) to check that your `SKILL.md` frontmatter follows the Agent Skills specification naming and format conventions.
</Tip>

## Backends and remote skill loading

The [Get started](#get-started) example uses `FilesystemBackend`, which loads skills from disk. Deep Agents supports other backends depending on how you want to store and manage skill files:

:::python
<SkillsUsageTabsPy />
:::

:::js
<SkillsUsageTabsJs />
:::

> <ParamField body="skills" type="list[str]" optional>

Reviewer: this probably now needs to move to get started. And we probably want to call that `Usage`. It should be clear to find.

For more information on subagent configuration and skills inheritance, see [Subagents](/oss/deepagents/subagents).

> ## Skill permissions

Reviewer: I wonder whether we want to frame this more about sharing skills? That woudl pick up the reusability also once more which might be good? Just a thought, permissions might be fine
